fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/HNSW.java

```diff
@@ -0,0 +1,2022 @@
+     * <p>
+     * For each {@link NodeReference} in the provided iterable, this method concurrently fetches the corresponding
+     * {@code Node} using the given {@link StorageAdapter}. The logic delegates to
+     * {@code fetchNodeIfNecessaryAndApply}, which determines whether a full node fetch is required.
+     * If a node is fetched from storage, the {@code biMapFunction} is applied. If the fetch is bypassed
+     * (e.g., because the reference itself contains sufficient information), the {@code fetchBypassFunction} is used
+     * instead.
+     *
+     * @param <R> The type of the node references to be processed, extending {@link NodeReference}.
+     * @param <N> The type of the key references within the nodes, extending {@link NodeReference}.
+     * @param <U> The type of the result after applying one of the mapping functions.
+     * @param storageAdapter The {@link StorageAdapter} used to fetch nodes from the underlying storage.
+     * @param readTransaction The {@link ReadTransaction} context for the read operations.
+     * @param storageTransform an affine transformation operator that is used to transform the fetched vector into the
+     *        storage space that is currently being used
+     * @param layer The layer index from which the nodes are being fetched.
+     * @param nodeReferences An {@link Iterable} of {@link NodeReference}s for the nodes to be fetched and processed.
+     * @param fetchBypassFunction The function to apply to a node reference when the actual node fetch is bypassed,
+     * mapping the reference directly to a result of type {@code U}.
+     * @param biMapFunction The function to apply when a node is successfully fetched, mapping the original
+     * reference and the fetched {@link Node} to a result of type {@code U}.
+     *
+     * @return A {@link CompletableFuture} that, upon completion, will hold a {@link java.util.List} of results
+     * of type {@code U}, corresponding to each processed node reference.
+     */
+    @Nonnull
+    private <R extends NodeReference, N extends NodeReference, U> CompletableFuture<List<U>>
+            fetchSomeNodesAndApply(@Nonnull final StorageAdapter<N> storageAdapter,
+                                   @Nonnull final ReadTransaction readTransaction,
+                                   @Nonnull final AffineOperator storageTransform,
+                                   final int layer,
+                                   @Nonnull final Iterable<R> nodeReferences,
+                                   @Nonnull final Function<R, U> fetchBypassFunction,
+                                   @Nonnull final BiFunction<R, Node<N>, U> biMapFunction) {
+        return forEach(nodeReferences,
+                currentNeighborReference -> fetchNodeIfNecessaryAndApply(storageAdapter, readTransaction,
+                        storageTransform, layer, currentNeighborReference, fetchBypassFunction, biMapFunction),
+                MAX_CONCURRENT_NODE_READS,
```

Comment:
Should this be a configuration option stead of a constant?

Reply:
done.